apps/web-client/src/lib/db/cr-sqlite/stock_cache.ts

```diff
@@ -0,0 +1,95 @@
+import { writable, get, derived } from "svelte/store";
+
+import type { DB, GetStockResponseItem } from "$lib/db/cr-sqlite/types";
+
+import { getStock } from "$lib/db/cr-sqlite/stock";
+import { reduce, wrapIter } from "@librocco/shared";
+
+const dbStore = writable<DB | null>(null);
+const valid = writable(false);
+
+/**
+ * Executes the stock query and sets the cache as valid (upon resolution)
+ */
+const execQuery = async (db: DB) => {
+	const stock = await getStock(db);
+	valid.set(true);
+	return stock;
+};
+
+/**
+ * An internal store keeping the full stock query as a promise
+ * NOTE: This is somewhat lazy - the initial promise never resolves, but it doesn't
+ * choke up the DB either. Only when the cached stock is activated (needed by a consumer), does it
+ * run the query.
+ *
+ * This is a tradeoff between prefetching the results and not blocking other DB interactions until the stock is needed.
+ */
+const query = writable<Promise<GetStockResponseItem[]>>(new Promise(() => {}));
+
+type StockByWarehouseMap = Map<number, Iterable<GetStockResponseItem>>;
+/**
+ * A store derived from cached stock query:
+ * - it contains a promise which resolves to a Map { warehouseId => Iterable<GetStockResponseItem> }
+ * - having a promise allows us to use Svelte's async await block
+ * - being a store, it automatically updates when the cache is invalidated
+ */
+export const stockByWarehouse = derived(
+	query,
+	($query) => $query.then((stock) => wrapIter(stock)._groupIntoMap((item) => [item.warehouseId, item])) as Promise<StockByWarehouseMap>
+);
+
+/**
+ * A store derived from cached stock query:
+ * - it contains a promise which resolves to a Map { warehouseId => number }
+ * - having a promise allows us to use Svelte's async await block
+ * - being a store, it automatically updates when the cache is invalidated
+ */
+export const warehouseTotals = derived(stockByWarehouse, ($stockByWarehouse) =>
+	$stockByWarehouse.then(
+		(stock) =>
+			// Reduce the quantities of items for each warehouse and create a Map { warehouseId => totalQuantiy }
+			new Map(wrapIter(stock).map(([warehouseId, items]) => [warehouseId, reduce(items, (acc, { quantity }) => acc + quantity, 0)]))
+	)
+);
+
+export const enable = (db: DB) => {
+	// Set the DB -- effectively enabling the cache
+	dbStore.set(db);
+
+	// If cache invalidated while not active, rerun the query
+	if (!get(valid)) {
+		query.set(execQuery(db));
+	}
+};
+
+export const disable = () => dbStore.set(null);
```

Comment:
Related to the comment on `valid`: shouldn't we also set `valid` to `false` here given the current architecture?

Reply:
No, it's a matter of keeping the caching logic as lazy as possible.

Consider this case:
- view A uses the cached data, view B doesn't 
- we switch from A -> B (cache is disabled)
- we switch to view C (which also uses cached data), it re-enables the cache and uses the cached value

Now, consider this:
- same setup: A, C - use the cache, B doesn't
- we're on view A, the cache gets invalidated
- since A uses the cache, we execute the (expensive) query to recalculate the state
- we switch from A -> B (cache is disabled)
- while on view B the cache gets invalidated, the query is not executed as it's expensive and unnecessary
- we switch to view C - we see that the cache is invalid, so we execute the expensive query